Allow configuration injection

Cendrao opened the pull request:
Following the [library guidelines](https://hexdocs.pm/elixir/master/library-guidelines.html) one of the best way to config a library is to allow the functions to inject their configurations. It gives a lot more flexibility to whoever uses it.

Resolves #14

Commit:
Allow configuration injection

Following the [library
guidelines](https://hexdocs.pm/elixir/master/library-guidelines.html)
one of the best way of config a library is allow the functions to inject
their configurations. It gives a lot more flexibility to whoever use it.

paulo-silva commented on lib/api_client.ex:
```diff
@@ -3,37 +3,37 @@ defmodule GlassFactoryApi.ApiClient do
       iex> GlassFactoryApi.ApiClient.get("members")
       {:ok , %{body: "[]", headers: [], status_code: 200}}
+
+  If no configuration is passed, it will get the default configuration which is defined in `GlassFacotryApi.Configuration.default_configuration`
```
```suggestion
  If no configuration is passed, it will get the default configuration which is defined in `GlassFactoryApi.Configuration.default_configuration`
```

paulo-silva commented on lib/core/configuration.ex:
```diff
@@ -0,0 +1,29 @@
+  @access_info Application.get_env(:glass_factory_api, GlassFactoryApi.ApiClient)
+
+  def default_configuration do
+    %Configuration{
```
Could we use pattern matching here to get values from `@access_info`? 🤔 

```elixir
%Configuration{
  subdomain: subdomain,
  user_token: user_token,
  user_email: user_email,
  api_url: api_url,
  adapter: GlassFactoryApi.TeslaAdapter
} = @access_info
```

(I'm not sure if `@access_info` has adapter key as well)

britto reviewed (commented):
Now that we are using Tesla, and it provides test mocks for us, we might no longer need to define our own adapter interface. Tesla itself can probably play that tole.

Cendrao commented on lib/members.ex:
```diff
@@ -23,8 +23,8 @@ defmodule GlassFactoryApi.Members do
       ]
   """
   @spec list_members() :: list(Member.t())
```
I still need to change this spec

Cendrao commented on lib/members.ex:
```diff
@@ -48,8 +48,8 @@ defmodule GlassFactoryApi.Members do
       }
   """
   @spec get_member(String.t()) :: Member.t() | nil
```
I still need to change this spec

britto commented on lib/core/configuration.ex:
```diff
@@ -0,0 +1,29 @@
+
+  defstruct [:subdomain, :user_token, :user_email, :api_url, :adapter]
+
+  @access_info Application.get_env(:glass_factory_api, GlassFactoryApi.ApiClient)
```
Here we're still relying on Application config, if the library users don't inject any config straight in the function calls and don't define these values in their mix config, we will fail without providing a helpful message.

My recommendation here would be making a clear fallback strategy, with everything evaluated at runtime (not compile time): first try function, then app config, and finally env var, if they all fail we raise something.

Cendrao committed:
Update lib/api_client.ex

Co-Authored-By: Paulo Henrique da Silva <paulo_silvalearn@hotmail.com>

Commit:
Change the ApiClient and Configuration module to build the configs

It also remove the HttpAdapter layer, as we are using Tesla, we can let
them handle the interface between clients HTTP.

Cendrao commented on mix.exs:
```diff
@@ -5,7 +5,7 @@ defmodule GlassFactoryApi.MixProject do
       app: :glass_factory_api,
       version: "0.1.0",
-      elixir: "~> 1.8",
+      elixir: "~> 1.9",
```
Now we need elixir 1.9+ because of `System.fetch_env!\1` used [here](https://github.com/Cendrao/glass_factory_api/pull/17/files#diff-610122b96b2e8892b9b99742443ff1fbR21)

Cendrao commented:
After a pair programming with @britto some big changes were made:

- `HttpAdapter` was removed, as we are using Tesla do handle HTTP client, let's permit they do it, so we don't need to interface with that anymore;
- We now try parsing configuration in 3 ways, [directly in the function](https://github.com/Cendrao/glass_factory_api/pull/17/files#diff-5249b954f1a72a4ac46ec0ba42414179R26) as in `Members.list_member\1` or in the `Configuration.build\1` which receives a map and try to evaluate its keys. If one of them is not on the map it looks in the Application variables and finally in Environment variables, if it's still not finding the needed config it raises an error ([look here](https://github.com/Cendrao/glass_factory_api/pull/17/files#diff-610122b96b2e8892b9b99742443ff1fbR6-R22));
- Taking advantage of `Tesla.Middleware`, [so our code looks way more elegant](https://github.com/Cendrao/glass_factory_api/pull/17/files#diff-6a0b0101a105cc5b6e013962e6c8e1e5R24-R28);

Commit:
Change elixir version in CircleCI

Commit:
Update docs

britto commented on config/dev.exs:
```diff
@@ -1,3 +1 @@
 use Mix.Config
```
Now we can delete the whole `config` fold altogether 👍

britto commented on lib/members.ex:
```diff
@@ -22,12 +22,11 @@ defmodule GlassFactoryApi.Members do
 
     body
-    |> Jason.decode!()
     |> Enum.map(&Member.to_struct(&1))
```
🎨 

```elixir
|> Enum.map(&Member.to_struct/1)
```

britto commented on lib/members.ex:
```diff
@@ -47,15 +46,14 @@ defmodule GlassFactoryApi.Members do
+      {:ok, %{status: 200, body: body}} ->
         body
-        |> Jason.decode!()
         |> Member.to_struct()
```
Most style guides recommend avoiding pipelines for a single call:

* https://github.com/christopheradams/elixir_style_guide#avoid-single-pipelines
* https://github.com/lexmag/elixir-style-guide#needless-pipeline

britto reviewed (commented):
This turned out great! 👍
